data/campaigns/tutorial/scenarios/02_Tutorial_part_2.cfg

```diff
@@ -484,6 +484,36 @@ A full list of abilities and weapons specials, along with traits, may be found i
         )}
     [/event]
 
+    # If the player didn't provide bait for Dumbo, the guardian AI won't move
+    # him, and that makes the scenario much harder than it should be. Hardcode
+    # the exact move that we want rather than switching to the normal AI, as
+    # the normal AI would probably take the village instead.
+    [event]
+        name=side 2 turn 2 end
+
+        [if]
+            [have_unit]
+                id=Dumbo
+                x,y=9,13
+            [/have_unit]
+            [not]
+                # Sanity check, if there's a unit anywhere here then the Guardian AI
+                # is expected to have attacked it; if not then something complex is
+                # happening (maybe the player already has a Ranger with Ambush).
+                [have_unit]
+                    x=8-10
+                    y=7-12
+                [/have_unit]
```

Comment:
The sanity check misses some hexes: 10,6 , 11,7 through 11,11 , and 12,9 , 12,10 (quick Archer), 12,11 (Konrad, if he didn't recruit). The grunt had the `quick` trait.

How about simplifying the check to `moves == max_moves && attacks_left=1`?

Reply:
The sanity check isn't to see if the guardian AI has moved the unit, it's to check that moving from 9,13 to 9,8 doesn't walk through a ZoC. We're in a side-end-turn event and Dumbo's still on his starting point, so the guardian AI has already decided not to move him (or a unit managed to get to 9,12 for a fight).

I didn't use `moves == max_moves` because the note about the AI changing `moves` in https://wiki.wesnoth.org/SingleUnitWML#Unit_State

If it sounds OK, then I'll just update the comment to say what it's checking.

Reply:
> I didn't use `moves == max_moves` because the note about the AI changing `moves` in https://wiki.wesnoth.org/SingleUnitWML#Unit_State

That says the AI may change `moves` _during_ the turn. I would guess that by the time the `end turn` event fires, the unit state will have been restored. But that's just a guess.

> If it sounds OK, then I'll just update the comment to say what it's checking.

Yes please. And by the way, I think the `y` range should start from 8, not 7. A unit on 8,7 / 9,7 / 10,7 shouldn't block the move.

Question. If the sanity check does _not_ hold, should we print a warning message? We're pretty unlikely to find out about the problem otherwise. (The tutorial uses a lot of custom code, but the code should be robust, since the tutorial is aimed at players with no experience at all.)

Reply:
Added a warning if he doesn't move over the bridge, unless he ends up next to a unit of side 1 (I assume the player has rushed him).

Not sure what to do about the player rushing him - maybe the north bridge should be 2 hexes wide?

Reply:
> Not sure what to do about the player rushing him - maybe the north bridge should be 2 hexes wide?

```
[event]
    name=moveto
    ...
    [message]
        speaker=Galdrad
        message=_ "Patience, young padawan! Do not recklessly charge ahead; wait for the other recruited units to join you and support your attack."
    [/message]
[/event]
```

Maybe also say something about potential traps, but that'd be easier if the enemy had units with ambush or nightstalk.

Reply:
I think we should give the player a 100% forced chance to hit if Dumbo is south of where he should be.

Rushing forwards seems reasonable to me - we're telling the player that they need to take the island quickly, and the other recruited units could be right behind and ready to support the attack. The grunt is so outnumbered that he shouldn't cross the bridge; it's a bad move even if he gets a free attack on the bait.

Reply:
> I would guess that by the time the end turn event fires, the unit state will have been restored. But that's just a guess

this is not the case, in fact the current behviour is quite bad and can cause other types of error aswell, see https://github.com/wesnoth/wesnoth/issues/4428

Reply:
> Rushing forwards seems reasonable to me - we're telling the player that they need to take the island quickly, and the other recruited units could be right behind and ready to support the attack.

Good point. Okay, how about having Dumbo on the north bank, heavily injured? Galdrad could advise Konrad to rush Dumbo with one or two units and in parallel send another unit, backed by a Shaman, to hold the island forest hex, (10,14), for ZoC purposes.

Reply:
That sounds like it would need checks of whether the player could kill Dumbo on turn 2, and the corresponding changes to turn 3's start event. I want to avoid having this PR grow to become a redesign.

Reply:
> I want to avoid having this PR grow to become a redesign

Absolutely!

You said you had a concern about what will happen if the player rushes Dumbo. What's the most difficult outcome? The player can't get a unit into shallow water on turn 1, so the player will suffer a 9x2 attack on 40% CTH, and then have to attack from the water..but I suppose they might miss and then get a 7x2 attack on 80% CTH on the next turn.

Anyway, none of this blocks merging, does it?

Reply:
I think the most difficult outcome is that the player gets delayed in getting on to the island, there's enough units and healing that I think it's all about position.

Thanks for the hint about merging, I'll do that now and may open another PR with ideas about the orc leader shouting at Dumbo to rush and attack, partly as a hint about how aggressive the rest of the orcs will be.